nvme_gw_persistence.py

```diff
@@ -0,0 +1,404 @@
+                                    (str(self.version),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+                self.logger.info(
+                    f"First gateway: created object {self.omap_name}")
+        except rados.ObjectExists:
+            self.logger.info(f"{self.omap_name} omap object already exists.")
+
+    def _write_key(self, key: str, val: str):
+        """Writes key and value to the persistent config."""
+
+        try:
+            version_update = int(self.version) + 1
+            with rados.WriteOpCtx() as write_op:
+                # Compare operation failure will cause write failure
+                write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                                  rados.LIBRADOS_CMPXATTR_OP_EQ)
+                self.ioctx.set_omap(write_op, (key,), (val,))
+                self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                    (str(version_update),))
+                self.ioctx.operate_write_op(write_op, self.omap_name)
+            self.version = version_update
+            self.logger.debug(f"omap_key generated: {key}")
+        except Exception as ex:
+            self.logger.error(f"Unable to write to omap: {ex}. Exiting!")
+            raise
+
+    def _delete_key(self, key: str):
+        """Deletes key from omap persistent config."""
+
+        version_update = int(self.version) + 1
+        with rados.WriteOpCtx() as write_op:
+            # Compare operation failure will cause delete failure
+            write_op.omap_cmp(self.OMAP_VERSION_KEY, str(self.version),
+                              rados.LIBRADOS_CMPXATTR_OP_EQ)
+            self.ioctx.remove_omap_keys(write_op, (key,))
+            self.ioctx.set_omap(write_op, (self.OMAP_VERSION_KEY,),
+                                (str(version_update),))
+            self.ioctx.operate_write_op(write_op, self.omap_name)
+        self.version = version_update
+        self.logger.debug(f"omap_key deleted: {key}")
+
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._write_key(key, val)
+
+    def delete_bdev(self, bdev_name: str):
+        """Deletes a bdev from the persistent config."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._delete_key(key)
+
+    def _restore_bdevs(self, omap_dict, callback):
+        """Restores a bdev from the persistent config."""
+
+        for (key, val) in omap_dict.items():
+            if key.startswith(self.BDEV_PREFIX):
+                args = self._clean_args(str(val, 'utf-8'))
+                req = pb2.bdev_create_req(
+                    bdev_name=args["bdev_name"],
+                    ceph_pool_name=args["ceph_pool_name"],
+                    rbd_name=args["rbd_name"],
+                    block_size=int(args["block_size"]),
+                )
+                ret = callback(req)
```

Comment:
nit: why do we need this `ret =` thing? It is not used and I expect the callback will rise exception if an error occurs?

BTW (probably could be improved in further PR), do we want to abort (as it is now) when restoring one bdev failed? E.g. it could be due to the image was accidentally deleted. Don't we want to just output an error message and proceed with other devices?

Reply:
To your first point, you are correct that it's not needed. I will remove it.
To your second, I'm open to discussion. My thought in writing it to abort was for a multi-gateway case, where all the gateways in the group should have identical configuration.

Reply:
If the error is on ceph side (e.g. an image is deleted) then I expect all gateways in the group will die. Anyway, sure, it is something that may be improved later. I just noticed it as a potential issue we will need to resolve.

Reply:
What do you think about using Github Issues to track these ideas? I'd like to have one place where we keep track of these potential issues / current issues for later PRs and we can assign ourselves as we work on them.